### PR TITLE
[compile] fix diagonal_scatter backward

### LINKED ISSUE #183602: torch.compile produces wrong gradient for diagonal_scatter backward
### 🐛 Describe the bug

`torch.compile` produces an incorrect gradient for `torch.diagonal_scatter` in backward.

Eager execution gives the expected non-zero gradient for `x`, while the compiled version produces a zero gradient.

```
import torch

def fn(x, src):
    return torch.diagonal_scatter(x, src, 0).sum()

x1 = torch.zeros(5, 5, requires_grad=True)
s1 = torch.ones(5, requires_grad=True)
fn(x1, s1).backward()

torch._dynamo.reset()

x2 = torch.zeros(5, 5, requires_grad=True)
s2 = torch.ones(5, requires_grad=True)
torch.compile(fn)(x2, s2).backward()

print("eager x.grad sum:", float(x1.grad.sum()))
print("compile x.grad sum:", float(x2.grad.sum()))
```

### Error logs

```
eager x.grad sum: 20.0
compile x.grad sum: 0.0
```

### Versions

PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Laptop GPU
Nvidia driver version: 545.92
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] onnx==1.21.0
[pip3] onnx2torch==1.5.15
[pip3] onnxruntime==1.23.2
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[pip3] triton==3.6.0

cc @pragupta @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by khushi-411**
Fixes: #183602

Now this works:
```python
In [1]: import torch
   ...: 
   ...: def fn(x, src):
   ...:     return torch.diagonal_scatter(x, src, 0).sum()
   ...: 
   ...: x1 = torch.zeros(5, 5, requires_grad=True)
   ...: s1 = torch.ones(5, requires_grad=True)
   ...: fn(x1, s1).backward()
   ...: 
   ...: torch._dynamo.reset()
   ...: 
   ...: x2 = torch.zeros(5, 5, requires_grad=True)
   ...: s2 = torch.ones(5, requires_grad=True)
   ...: torch.compile(fn)(x2, s2).backward()
   ...: 
   ...: print("eager x.grad sum:", float(x1.grad.sum()))
   ...: print("compile x.grad sum:", float(x2.grad.sum()))

eager x.grad sum: 20.0
compile x.grad sum: 20.0
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @xmfan @aditvenk

### CODE REVIEW

**Comment by Skylion007** [comment]
This still works with nonzero offset?? Or non default args?

**Comment by khushi-411** [comment]
Hi, @Skylion007! Yes, it does.
```python

In [4]: import torch
   ...: 
   ...: def fn(x, a):
   ...:     return torch.diagonal_scatter(x, a, offset=1, dim1=0, dim2=2).sum()
   ...: 
   ...: x1 = torch.zeros(5, 4, 5, requires_grad=True)
   ...: s1 = torch.ones(4, 4, requires_grad=True)
   ...: fn(x1, s1).backward()
   ...: 
   ...: torch._dynamo.reset()
   ...: 
   ...: x2 = torch.zeros(5, 4, 5, requires_grad=True)
   ...: s2 = torch.ones(4, 4, requires_grad=True)
   ...: torch.compile(fn)(x2, s2).backward()
   ...: 
   ...: print("eager   x.grad sum:", float(x1.grad.sum()))
   ...: print("compile x.grad sum:", float(x2.grad.sum()))
   ...: print("grads match:", torch.allclose(x1.grad, x2.grad))

eager   x.grad sum: 84.0
compile x.grad sum: 84.0
grads match: True
```

**Comment by jansel** [comment]
@claude review these changes

**Comment by khushi-411** [comment]
Hi, @jansel, @Skylion007! Gentle ping, the CI is green. Could you please review the PR?

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -9849,6 +9849,36 @@ def fn(x, a):
...
+        cases = [
+            ((5, 5), (5,), 0, 0, 1),
+            ((5, 4, 5), (4, 4), 1, 0, 2),
+            ((3, 5, 5), (3, 4), 1, 1, 2),
+            ((5, 5), (4,), -1, 0, 1),
+        ]
+
+        compiled_fn = torch.compile(fn, backend="inductor", fullgraph=True)
+
+        for x_shape, y_shape, offset, dim1, dim2 in cases:
```
Comment (Skylion007) [inline test/inductor/test_torchinductor.py:10088]: Slight nit: would be better with self.subTest with context manager to diagnose fails easier in the future.

Comment (khushi-411) [reply test/inductor/test_torchinductor.py:10088]: Nice catch, thanks! Updated the loop to use `self.subTest` as a context manager.

**Review by jansel — DISMISSED** [review]

**Comment by khushi-411** [comment]
@pytorchbot merge

Thanks for reviewing the PR, @Skylion007 and @jansel! :)

**Review by Skylion007 — DISMISSED** [review]

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by khushi-411** [comment]
Hi, my reason for adding the xfail in test_dtensor_ops.py and removing it from common_ops_unbacked.py was that when I removed `diagonal_scatter` from `torch/testing/_internal/common_ops_unbacked.py` I got:
[Link](https://github.com/pytorch/pytorch/actions/runs/25870487107/job/76028172781)
```python
==================================== RERUNS ====================================
_____ TestOpsUnbackedCPU.test_unbacked_op_db_diagonal_scatter_cpu_float32 ______
Unexpected success
_____ TestOpsUnbackedCPU.test_unbacked_op_db_diagonal_scatter_cpu_float32 ______
Unexpected success
=================================== FAILURES ===================================
_____ TestOpsUnbackedCPU.test_unbacked_op_db_diagonal_scatter_cpu_float32 ______
Unexpected success
```
The unbacked test was no longer failing. However, DTensor tests were still failing, specifically:
```python
The failing test is TestUnbackedDTensorOpsCPU::test_unbacked_dtensor_op_db_diagonal_scatter_cpu_float32
```
So I added an `xfail` in `test/distributed/tensor/test_dtensor_ops.py` for the DTensor-specific failure.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by khushi-411** [comment]
closing this pr, because everyone has subscribed to it, will create a new pr. Thanks!
